keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,997 @@
+I1026 16:21:00.461886  801139 scheduler.go:464] "Status after running PostFilter plugins for pod" pod="default/my-csi-app-inline-volume" status=&{code:2 reasons:[0/1 nodes are available: 1 Preemption is not helpful for scheduling.] err:<nil> failedPlugin:}
+I1026 16:21:00.461918  801139 factory.go:209] "Unable to schedule pod; no fit; waiting" pod="default/my-csi-app-inline-volume" err="0/1 nodes are available: 1 node(s) did not have enough free storage."
+```
+
+### Test Plan
+
+The new code will be covered by unit tests that execute as part of `pull-kubernetes-unit`.
+
+Converted components will be tested by exercising them with JSON output at high
+log levels to emit as many log messages as possible. Analysis of those logs
+will detect duplicate keys that might occur when a caller uses `WithValues` and
+a callee adds the same value in a log message.
+
+Static code analysis cannot detect this, or at least not easily. What it can
+check is that the individual log calls pass valid key/value pairs.
+
+
+### Graduation Criteria
+
+#### Alpha
+
+- Common utility code available (logcheck with the additional checks,
+  `k8s.io/klogr` v0.1.0)
+- Documentation for developers available at https://github.com/kubernetes/community
+- At least kube-scheduler framework and some plugins converted
+- Initial e2e tests completed and enabled in Prow
+
+#### Beta
+
+- All of kube-scheduler (in-tree) and CSI external-provisioner (out-of-tree) converted
+- Users of klog notified through a Kubernetes blog post and an email to
+  dev@kubernetes.io that a logger must be set with k8s.io/klogr.
+- Issues filed against kubernetes-sigs projects with a request to set a logger
+  with k8s.io/klogr. These issues will have a link to the enhancement issue
+  (for cross-referencing) and a link to the developer documentation.
+
+#### GA
+
+- All code in kubernetes/kubernetes converted to contextual logging,
+  no dependency on klog anymore
+- All code under kubernetes-sigs sets a logger with k8s.io/klogr
+  and therefore is ready for the next k/k release.
+- Gather feedback from developers and surveys
+- Allowing time for feedback
+
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+
+### Upgrade / Downgrade Strategy
+
+Not applicable. The log output will be determined by what is implemented in the
+code that currently runs.
+
+### Version Skew Strategy
+
+Not applicable. The log output format is the same as before, therefore other
+components are not affected.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+This is not a feature in the traditional sense. Converted log calls will always
+use contextual logging, the only supported rollback is to downgrade to a
+previous release.
```

Comment:
I'm not sure how I feel about this. What happens if this causes a blocking performance degradation in the scheduler? The only option is to downgrade it a major version? Is it completely impossible to provide an escape valve?

Reply:
We could have a feature gate that does several things:
- controls usage of WithName/WithValues
- simplifies klogr so that it uses a global logger instead of checking the context

I fear that this will make the code very ugly and potentially buggy. In particular conditional usage of WithName/WithValues will need additional variables outside of the if statement - I tried that with the scheduler code depending on the log verbosity and it's really not nice.

Not doing the context lookup doesn't save much based on what we have seen so far. Having the if check just makes it (slightly) worse.

Reply:
I wonder if we could have a feature flag that, when disabled, makes klogr use the global logger, and when enabled, uses the specified logger... would that work?

Reply:
This would currently only help with `klogr.FromContext`, which doesn't seem to be causing any noticeable overhead so far.

`klogr` doesn't control what components do with `WithValue` and `WithName` or `context.WithValue` and therefore cannot shortcut those. We would have to introduce a new set of APIs for all of those:

```
func LoggerWithValues(logger Logger, kv ...interface{}) Logger {
   if contextualEnabled {
      return logger.WithValues(kv...)
   }
   return logger
}

func LoggerWithName(logger Logger, name string) Logger {
    if contextualEnabled {
      return logger.WithName(name)
    }
    return logger
}

ContextWithLogger(ctx context, logger Logger) ctx {
    if contextualEnabled {
       return NewContext(ctx, logger)
    }
    return ctx
}
```

Connecting this `contextualEnabled` with some feature gate would have to be done by the `klogr` user. The package itself cannot depend on the packages in Kubernetes which handle feature gates.

This is doable, but so far it doesn't seem to be particularly useful. Please let me know if I shall add it.

None of that helps for the overhead that comes from adding `ctx` or `logger` parameters that otherwise wouldn't be there.

Reply:
I managed to convince myself that such a feature gate is doable and will add it first to the prototype, then the KEP.

Reply:
Excellent. Feature gates are the standard way, let's do that and then I think it satisfies PRR requirements.